src/ucp/core/ucp_ep.c

```diff
@@ -232,55 +232,187 @@ ucp_ep_adjust_params(ucp_ep_h ep, const ucp_ep_params_t *params)
+        goto out_free_address;
+    }
+
+    status = ucp_ep_adjust_params(ep, params);
+    if (status != UCS_OK) {
+        ucp_ep_destroy_internal(ep);
+        goto out_free_address;
+    }
+
+    ep->conn_sn = conn_sn;
+
+    /*
+     * If we are connecting to our own worker, and loopback is allowed, connect
+     * the endpoint to itself by updating dest_ep_ptr.
+     * Otherwise, add the new ep to the matching context as an expected endpoint,
+     * waiting for connection request from the peer endpoint
+     */
+    flags = UCP_PARAM_VALUE(EP, params, flags, FLAGS, 0);
+    if ((remote_address.uuid == worker->uuid) &&
+        !(flags & UCP_EP_PARAMS_FLAGS_NO_LOOPBACK)) {
+        ucp_ep_update_dest_ep_ptr(ep, (uintptr_t)ep);
+        ucp_ep_flush_state_reset(ep);
+    } else {
+        ucp_ep_match_insert_exp(&worker->ep_match_ctx, remote_address.uuid, ep);
+    }
+
+    /* if needed, send initial wireup message */
+    if (!(ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED)) {
+        ucs_assert(!(ep->flags & UCP_EP_FLAG_CONNECT_REQ_QUEUED));
+        status = ucp_wireup_send_request(ep);
         if (status != UCS_OK) {
-            goto err_free_address_list;
+            goto out_free_address;
         }
+    }
 
-        ucs_free(local_address.address_list);
-        ucs_free(address_buffer);
+    status = UCS_OK;
+
+out_free_address:
+    ucs_free(remote_address.address_list);
+out:
+    if (status == UCS_OK) {
+        *ep_p = ep;
     }
+    return status;
+}
 
-    return UCS_OK;
+ucs_status_t ucp_worker_create_loopback_ep(ucp_worker_h worker,
+                                           uint64_t tl_bitmap,
+                                           unsigned ep_init_flags,
+                                           const char *message, ucp_ep_h *ep_p)
+{
+    ucp_ep_params_t ep_params;
+    ucp_unpacked_address_t local_address;
+    void *address_buffer;
+    size_t address_length;
+    int need_wireup;
+    ucp_rsc_index_t rsc_index;
+    ucs_status_t status;
 
-err_free_address_list:
-    ucs_free(local_address.address_list);
-err_free_address_buffer:
-    ucs_free(address_buffer);
-err_cleanup_eps:
-    for (i = 0; i < UCS_MEMORY_TYPE_LAST; i++) {
-        if (worker->mem_type_ep[i]) {
-           ucp_ep_destroy_internal(worker->mem_type_ep[i]);
+    status = ucp_address_pack(worker, NULL, tl_bitmap,
+                              UCP_ADDRESS_PACK_FLAG_ALL, NULL, &address_length,
+                              &address_buffer);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucp_address_unpack(worker, address_buffer,
+                                UCP_ADDRESS_PACK_FLAG_ALL, &local_address);
+    if (status != UCS_OK) {
+        goto free_address_buffer;
+    }
+
+    need_wireup = 0;
+    ucs_for_each_bit(rsc_index, tl_bitmap) {
+        need_wireup = ucp_worker_is_tl_p2p(worker, rsc_index);
+        if (need_wireup) {
+            break;
         }
     }
+
+    if (need_wireup) {
```

Comment:
why not use ucp_ep_create_api_to_worker_addr() always?
also, would not need to unpack address here

Reply:
mem_type EPs initially did not have wireup code at all, this function was added to unite address related code.
indeed, unpack can be done before ucp_ep_create_to_worker_addr

Reply:
so any reason not to use ucp_ep_create_api_to_worker_addr() now for both cases?

Reply:
not sure if it may cause some side effects aside from extra work on initialization.

Reply:
but you already use it for one case, so would be simpler to use it for both cases